### DISCUSSION TITLE
API route built as 'Static', forcing a serverside cache

### DISCUSSION THREAD

**Discussion by jschaf14** [vercel/community#2544] [Help] [answered]
### Summary

I am trying to deploy a next.js app on Vercel and am running into the issue of a 304 return when pinging the '/api/prompt' route which should return the most up to date results of user input prompts. I believe this is due to Vercel building the file as a static file and I cannot figure out how to have Vercel stop building it as a static file.

Build Image:
<img width="1200" alt="Screen Shot 2023-05-21 at 9 11 44 PM" src="https://github.com/vercel/community/assets/23328479/c679afb2-080d-468b-b67d-a1e639917be7">

Problem Line (Hypothesized):
`├ ○ /api/prompt                            0 B                0 B`

Code in the /api/prompt/route.js:
```
import { connectToDB } from "@utils/database";
import Prompt from "@models/prompt";

export const GET = async (request, { params }) => {
  try {
    await connectToDB();
    const prompts = await Prompt.find( {} )
      .populate('creator')
      .sort({ createdAt: 1 });

    return new Response(JSON.stringify(prompts), {
      status: 200,
      headers: {
        'cache-control': 's-maxage=1, stale-while-revalidate=59',
      }
    });
  } catch (error) {
    console.log(error);
    return new Response("Failed to fetch all prompts", {
      status: 500
    })
  }
}
```

I have tried many different kinds of ways to prevent caching but all of them have been ignored due to the statement: 
'Vercel does not allow bypassing the cache for static files by design.' [Vercel Docs](https://vercel.com/docs/concepts/edge-network/caching)

### Additional information

[Promptopia Github Repo](https://github.com/jschaf14/promptopia)

### Example
![Screen Shot 2023-05-21 at 9 55 22 PM](https://github.com/vercel/community/assets/23328479/cacc0d3f-fc9e-490d-b2ca-bac9dc8f75e6)

**Comment by leerob** [comment, marked as answer]
Route Handlers are static by default: https://nextjs.org/docs/app/building-your-application/routing/router-handlers#static-route-handlers

You can use `export const dynamic = 'force-dynamic'` if you never want caching.

**Reply by jschaf14** [reply]
Leerob, thank you so much for this answer. 

Quick note on your answer, I believe you meant `export const dynamic = 'force-dynamic'` instead of exporting runtime. Reference: [Route Segment Config](https://nextjs.org/docs/app/api-reference/file-conventions/route-segment-config).

This did fix my issue but I realize that it is not necessarily good practice. If you know where to direct me to learn how to make use of caching the response and re-validating it every so often to keep it up to date I would appreciate that. I attempted `export const revalidate = 60;` without success.

**Reply by leerob** [reply]
Whoops, yes, updated. Does this work?

```
import { connectToDB } from "@utils/database";
import Prompt from "@models/prompt";

export const revalidate = 60;

export const GET = async (request, { params }) => {
  try {
    await connectToDB();
    const prompts = await Prompt.find( {} )
      .populate('creator')
      .sort({ createdAt: 1 });

    return new Response(JSON.stringify(prompts));
  } catch (error) {
    console.log(error);
    return new Response("Failed to fetch all prompts", {
      status: 500
    })
  }
}
```

**Reply by leerob** [reply]
I'm guessing you might want to be streaming responses back here, so you might want the caching layer to happen on the OpenAI calls themselves, and not the entire response from the function, but not sure.
